tests/test_fluent.py

```diff
@@ -0,0 +1,411 @@
+# Copyright Mozilla Foundation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from textwrap import dedent
+from unittest import TestCase
+
+from fluent.syntax import ast as ftl
+
+from moz.l10n.fluent import fluent_parse, fluent_parse_message, fluent_serialize
+from moz.l10n.message import (
+    CatchallKey,
+    Expression,
+    FunctionAnnotation,
+    PatternMessage,
+    SelectMessage,
+    VariableRef,
+)
+from moz.l10n.resource import Comment, Entry, Metadata, Resource, Section
+
+# Show full diff in self.assertEqual. https://stackoverflow.com/a/61345284
+# __import__("sys").modules["unittest.util"]._MAX_LENGTH = 999999999
+
+
+class TestFluent(TestCase):
+    def test_fluent_value(self):
+        source = dedent(
+            """\
+            key =
+                pre { $a ->
+                    [1] One
+                   *[2] Two
+                } mid { $b ->
+                   *[bb] BB
+                    [cc] CC
+                } post
+                .attr = foo
+            """
+        )
+        res = fluent_parse(source)
+        self.assertEqual(len(res.sections), 1)
+        self.assertEqual(len(res.sections[0].entries), 2)
+        self.assertEqual(res.sections[0].entries[0].id, ["key"])
+        self.assertIsInstance(res.sections[0].entries[0].value, ftl.Pattern)
+        self.assertEqual(res.sections[0].entries[1].id, ["key", "attr"])
+        self.assertIsInstance(res.sections[0].entries[1].value, ftl.Pattern)
+        self.assertEqual("".join(fluent_serialize(res)), source)
+
+    def test_equality_same(self):
+        source = 'progress = Progress: { NUMBER($num, style: "percent") }.'
+        res1 = fluent_parse(source, fluent_parse_message)
+        res2 = fluent_parse(source, fluent_parse_message)
+        self.assertEqual(res1, res2)
+
+    def test_equality_different_whitespace(self):
+        source1 = b"foo = { $arg }"
+        source2 = b"foo = {    $arg    }"
+        res1 = fluent_parse(source1, fluent_parse_message)
+        res2 = fluent_parse(source2, fluent_parse_message)
+        self.assertEqual(res1, res2)
+
+    def test_resource(self):
```

Comment:
I inlined the FTL contents because `FluentParser.parse()` requires a `str`, and so there was no reason to test its `bytes` support as with some of the other formats.

Adding the `demo.ftl` here does seem like a good idea.